### DISCUSSION TITLE
How to include Python.h in MOOSE code?

### DISCUSSION THREAD

**Discussion by Wang-Yihu** [idaholab/moose#26507] [Q&A General] [answered]
I can easily include Python.h in gcc environment like this:
File runpy.cpp
```C++
#include "/usr/include/python3.10/Python.h"
int main()
{
  Py_Initialize();
  PyRun_SimpleString("print('Hello world!')");
  PyFinalize();
  return 0;
}
```
Input in shell:
```
gcc runpy.cpp -o runpy -I/usr/include/python3.10 -lpython3.10
./runpy
```
It will certainly return:
```
Hello world!
```
However, how to compile Python.h in MOOSE codes?
When I try to input include "/usr/include/python3.10/Python.h" in my source code. As I build
```
conda activate moose
make
```
The shell will return errors like:
```shell
Checking if header needs updating: /home/sjtu/projects/moose/framework/include/base/MooseRevision.h...
Creating Unity Directory /home/sjtu/projects/lianxi_couplewithpy/build/unity_src
Creating Unity /home/sjtu/projects/lianxi_couplewithpy/build/unity_src/problem_Unity.C
Rebuilding symlinks in /home/sjtu/projects/lianxi_couplewithpy/build/header_symlinks
Compiling C++ (in opt mode) /home/sjtu/projects/lianxi_couplewithpy/build/unity_src/problem_Unity.C...
In file included from /usr/include/python3.10/Python.h:8,
                 from /home/sjtu/projects/lianxi_couplewithpy/src/problem/ExternalZidingyiQiujieqiChuanshuShujuProblem.C:10,
                 from /home/sjtu/projects/lianxi_couplewithpy/build/unity_src/problem_Unity.C:2:
/usr/include/python3.10/pyconfig.h:3:12: fatal error: x86_64-linux-gnu/python3.10/pyconfig.h: No such file or directory
    3 | #  include <x86_64-linux-gnu/python3.10/pyconfig.h>
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [/home/sjtu/projects/moose/framework/build.mk:151: /home/sjtu/projects/lianxi_couplewithpy/build/unity_src/problem_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
```

I try to input ADDITIONAL_LIBS ="-I/usr/include/python3.10/ -lpython3.10", but it seems useless... It will return the same error...

**Comment by milljm** [comment]
What does the following return:
```bash
python3-config
```

For me, `python3-config` resides in my `moose` Conda environment. And so when I ask it what my include flags should be, I am provided with:
```bash
❯ python3-config --includes
-I/Users/milljm/conda3/envs/moose/include/python3.10 -I/Users/milljm/conda3/envs/moose/include/python3.10
```
I suppose that means the `python` interpreter I am using, should have its matching includes in the above location:
```bash
❯ file /Users/milljm/conda3/envs/moose/include/python3.10/python.h
/Users/milljm/conda3/envs/moose/include/python3.10/python.h: c program text, ASCII text
```

So perhaps the error you are receiving is mixing and matching different python includes (by hard coding a path to system python). Unless of course, you are actually wanting to use system python over the one supplied by the Conda moose environment?

**Reply by Wang-Yihu** [reply]
You are right. It seems that there are two pythons for base and conda moose system, respectively...
I input
```
python3-config
```
It will return:
```shell
Usage: /home/sjtu/mambaforge3/bin/python3-config –prefix|--exec-prefix|--includes|--libs|--cflags|--ldflags|--extension-suffix|--help|--abiflags|--configdir|--embed
```
In the base environment, I input
```shell
python3-config --includes
```
It will return
```
-I/home/sjtu/mambaforge3/include/python3.10 -I/home/sjtu/mambaforge3/include/python3.10
```
However in conda MOOSE environment, the return is different. I input
```
conda activate moose
python3-config --includes
```
It will return a different directory
```
-I/home/sjtu/mambaforge3/envs/moose/include/python3.10 -I/home/sjtu/mambaforge3/envs/moose/include/python3.10
```
It seems that I should use system python rather than Conda moose environment... So what should I do? em...

**Reply by milljm** [reply]
There a several Makefile's in MOOSE which call for a Python interpreter. I think if you took a look at how we are determining where/what python includes there are available in the following Makefile, it might make more sense:

https://github.com/idaholab/moose/blob/3a1efc798c7227ea5057590ab37e266f1201845b/framework/contrib/hit/Makefile

Edit:
Basically, instead of hard-coding a path to some version of Python, the logic in the above Makefile allows for code to use whatever python is currently being made available.

**Reply by Wang-Yihu** [reply]
Milljm:
Could you tell me how to use this `Makefile` for including `Python.h` please?
In fact, I have never thought about the compiling process for MOOSE before. For solving this problem, recently, I have read the `Makefile` in user's app directory. In this `Makefile`, three mk files- `build.mk`, `moose.mk` and `app.mk`(not considering `modules.mk`) have been included. I insert `echo` command in these three files to trace what will happen after I `make`. And I found the shell will do (the name of user's MOOSE App is lianxi_mf in this example)
```
/home/sjtu/projects/moose/framework/scripts/get_repo_revision.py /home/sjtu/projects/moose/framework /home/sjtu/projects/moose/framework/include/base/MooseRevision.h MOOSE
$$(shell mkdir -p /home/sjtu/projects/lianxi_mf/build/header_symlinks)
ln -sf /home/sjtu/projects/lianxi_mf/include/base/lianxi_mfApp.h /home/sjtu/projects/lianxi_mf/build/header_symlinks/lianxi_mfApp.h
ln -sf /home/sjtu/projects/lianxi_mf/test/include/base/lianxi_mfTestApp.h /home/sjtu/projects/lianxi_mf/build/header_symlinks/lianxi_mfTestApp.h
/home/sjtu/mambaforge3/envs/moose/libmesh/contrib/bin/libtool  ...(LOTS OF OPTIONS!)
/home/sjtu/mambaforge3/envs/moose/libmesh/contrib/bin/libtool  ...(LOTS OF OPTIONS!)
/home/sjtu/mambaforge3/envs/moose/libmesh/contrib/bin/libtool  ...(LOTS OF OPTIONS!)
/home/sjtu/mambaforge3/envs/moose/libmesh/contrib/bin/libtool  ...(LOTS OF OPTIONS!)
/home/sjtu/mambaforge3/envs/moose/libmesh/contrib/bin/libtool  ...(LOTS OF OPTIONS!)
/home/sjtu/mambaforge3/envs/moose/libmesh/contrib/bin/libtool  ...(LOTS OF OPTIONS!)
/home/sjtu/mambaforge3/envs/moose/libmesh/contrib/bin/libtool . ..(LOTS OF OPTIONS!)
some thing about  /scripts/write_appresource_file.py
/home/sjtu/mambaforge3/envs/moose/libmesh/contrib/bin/libtool . ..(LOTS OF OPTIONS!)
```
It seems that this `Makefile`(in the directory `moose/framework/contrib/hit`
) for Python interpreter you mentioned 
https://github.com/idaholab/moose/blob/3a1efc798c7227ea5057590ab37e266f1201845b/framework/contrib/hit/Makefile
have not work in the default MOOSE compiling process. 

**So how to use this `Makefile` for Python.h? Is there any tutorials?**

I hope to receive your reply soon!
Sincerely yours
WANG Yihu

**Reply by Wang-Yihu** [reply]
Jason Miller：
In fact, I do not understand _the logic in the above Makefile_. I try to change this Makefile 
https://github.com/idaholab/moose/blob/3a1efc798c7227ea5057590ab37e266f1201845b/framework/contrib/hit/Makefile
like this:
1. Change the `pyconfig `into the result of `python3-config` in the base environment instead of conda moose environment.
2. Add `-I/home/sjtu/mambaforge3/include/python3.10` and `-I/home/sjtu/mambaforge3/include/python3.10` at the end of the shell command.
Now the Makefile in https://github.com/idaholab/moose/blob/3a1efc798c7227ea5057590ab37e266f1201845b/framework/contrib/hit/Makefile
is changed into:
```makefile
CXX ?= g++
UNAME := $(shell uname)

#pyconfig := python3-config
pyconfig :="Usage: /home/sjtu/mambaforge3/bin/python3-config --prefix|--exec-prefix|--includes|--libs|--cflags|--ldflags|--extension-suffix|--help|--abiflags|--configdir|--embed"
ifeq (, $(shell which $(pyconfig) 2>/dev/null))
	pyconfig := python-config
endif
cython := cython3
ifeq (, $(shell which $(cython) 2>/dev/null))
	cython := 'cython -3'
endif

ifeq ($(UNAME), Darwin)
	DYNAMIC_LOOKUP := -undefined dynamic_lookup
else
	DYNAMIC_LOOKUP :=
endif

$(info Building hit for python with $(pyconfig))

PYTHONPREFIX ?= `$(pyconfig) --prefix`
PYTHONCFLAGS ?= `$(pyconfig) --cflags`
HITCPP := hit.cpp

WASP_DIR           ?= $(abspath ../wasp/install)

# find the wasp libraries
lib_suffix := so
ifeq ($(shell uname -s),Darwin)
	lib_suffix := dylib
	wasp_LIBS         := $(shell find -E $(WASP_DIR)/lib -regex ".*/lib[a-z]+.$(lib_suffix)")
else
	wasp_LIBS         := $(wildcard $(WASP_DIR)/lib/libwasp*$(lib_suffix))
endif
wasp_LIBS         := $(notdir $(wasp_LIBS))
wasp_LIBS         := $(patsubst %.$(lib_suffix),%,$(wasp_LIBS))
wasp_LIBS         := $(patsubst lib%,-l%,$(wasp_LIBS))
ifeq ($(wasp_LIBS),)
  $(error WASP does not seem to be available. Make sure to either run scripts/update_and_rebuild_wasp.sh in your MOOSE directory, or set WASP_DIR to a valid WASP install)
endif

wasp_CXXFLAGS  += -I$(WASP_DIR)/include
wasp_LDFLAGS   += -Wl,-rpath,$(WASP_DIR)/lib -L$(WASP_DIR)/lib $(wasp_LIBS)

hit: main.cc parse.cc lex.cc braceexpr.cc braceexpr.h lex.h parse.h
	$(CXX) -std=c++17 $(wasp_CXXFLAGS) -g $(CXXFLAGS) $< parse.cc lex.cc braceexpr.cc -o $@ $(wasp_LDFLAGS) -I/usr/include/python3.10/ -lpython3.10

bindings: hit.so

hit.so: parse.cc lex.cc braceexpr.cc
	$(CXX) -std=c++17 $(wasp_CXXFLAGS) -w -fPIC -lstdc++ -shared -L$(PYTHONPREFIX)/lib $(PYTHONCFLAGS) $(DYNAMIC_LOOKUP) $^ $(HITCPP) -o $@ $(wasp_LDFLAGS) -I/usr/include/python3.10/ -lpython3.10

$(HITCPP): hit.pyx chit.pxd
	$(cython) -o $@ --cplus $< -I/usr/include/python3.10/ -lpython3.10

.PRECIOUS: $(HITCPP)

.PHONY: clean bindings

clean:
	rm -f hit hit.so
```
Now I input these 
```
cd /home/sjtu/projects/moose/framework/contrib/hit/
rm -r hit
touch *./
conda activate moose
make
```
Now a new hit is generated!
I also add `-I/home/sjtu/mambaforge3/include/python3.10` and `-I/home/sjtu/mambaforge3/include/python3.10` in `$(FRAMEWORK_DIR)/app.mk` and `$(FRAMEWORK_DIR)/build.mk` like this discussion.
However, when I try to build in my MOOSE app. Input 
```
cd home/sjtu/projects/lianxi_mf
conda activate moose
make
```
It failed again!
```
In file included from /usr/include/python3.10/Python.h:8,
                 from /home/sjtu/projects/lianxi_mf/src/main.C:16:
/usr/include/python3.10/pyconfig.h:3:12: fatal error: x86_64-linux-gnu/python3.10/pyconfig.h: No such file or directory
    3 | #  include <x86_64-linux-gnu/python3.10/pyconfig.h>
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [/home/sjtu/projects/moose/framework/build.mk:166: /home/sjtu/projects/lianxi_mf/src/main.x86_64-conda-linux-gnu.opt.lo] Error 1
```
It seems that I do not understand how MOOSE import python...

**Comment by Wang-Yihu** [comment, marked as answer]
Dear friends:
I have import `Python.h` in MOOSE code recently. Steps are shown below: (my username is `sjtu`)
1. Find head file `Python.h` and library file `libpython3.10.so` in the MOOSE environment(not base environment). These files' directories in my computer are `/home/sjtu/mambaforge3/envs/moose/include/python3.10/Python.h` and `/home/sjtu/mambaforge3/envs/moose/lib/libpython3.10.so`
2. Modify makefiles to add `-I`, `-L`, `-l` options. I add `-I/home/sjtu/mambaforge3/envs/moose/include/python3.10/ -L/home/sjtu/mambaforge3/envs/moose/lib/ -lpython3.10` to `app.mk` and `build.mk` directly as shown in https://github.com/idaholab/moose/discussions/26504
3. Input `#include "/home/sjtu/mambaforge3/envs/moose/include/python3.10/Python.h"` to your MOOSE C++ code.
4. Input these test code to your MOOSE C++ code.
```
Py_Initialize();
PyRun_SimpleString("print('Hello world!')");
PyFinalize();
```
This can be made in MOOSE environment in my computer successfully.
Certainly, this method is not perfect. It is not recommended to modify files in MOOSE framework  `app.mk` and `build.mk`. There might be environment variable for linking `Python.h` and  `libpython3.10.so` in the MOOSE environment instead of hard-coding a path.

**Reply by milljm** [reply]
You would want to modify your applications `app.mk` Makefile (and not the one in MOOSE) to link to what-ever-is you need from `python3-config --cflags`. I was only demonstrating how MOOSE makes use of some of Python's libraries by showing you the hit parser Makefile as an example, which makes use of `python3-config`.
